File: nextjs_app/lib/schemas/analysis.ts
import { z } from "zod";

export const analysisRunReadSchema = z.object({
  id: z.string(),
  comparison_group: z.string(),
  success_definition: z.record(z.string(), z.number()),
  total_analyzed: z.number().int(),
  total_excluded: z.number().int(),
  excluded_reasons: z.record(z.string(), z.number().int()),
  created_at: z.string(),
});

export const analysisRunRowReadSchema = z.object({
  id: z.string(),
  strategy_name: z.string(),
  strategy_version_id: z.string().nullable(),
  calmar_ratio: z.number().nullable(),
  sortino_ratio: z.number().nullable(),
  trades_per_year: z.number().nullable(),
  is_success: z.boolean(),
  indicators: z.array(z.string()),
  indicator_count: z.number().int(),
  parameter_count: z.number().int(),
  entry_archetype: z.string(),
  exit_archetype: z.string(),
  category: z.string().nullable(),
  direction: z.string().nullable(),
  mts_compatibility: z.string().nullable(),
});

export const cohortRowSchema = z.object({
  value: z.string(),
  success: z.number().int(),
  total: z.number().int(),
  success_quote: z.number().nullable(),
  lift: z.number().nullable(),
  median_calmar: z.number().nullable(),
});

export const analysisRunDetailReadSchema = analysisRunReadSchema.extend({
  rows: z.array(analysisRunRowReadSchema),
  cohort: z.array(cohortRowSchema),
});

export type AnalysisRunRead = z.infer<typeof analysisRunReadSchema>;
export type AnalysisRunRowRead = z.infer<typeof analysisRunRowReadSchema>;
export type CohortRow = z.infer<typeof cohortRowSchema>;
export type AnalysisRunDetailRead = z.infer<typeof analysisRunDetailReadSchema>;

export const AXIS_OPTIONS: { value: string; label: string }[] = [
  { value: "indicator", label: "Indikator" },
  { value: "indicator_count", label: "Indikatorzahl" },
  { value: "parameter_count", label: "Parameterzahl" },
  { value: "entry_archetype", label: "Entry-Archetyp" },
  { value: "exit_archetype", label: "Exit-Archetyp" },
  { value: "category", label: "Kategorie" },
  { value: "direction", label: "Richtung" },
  { value: "mts_compatibility", label: "MTS-Eignung" },
];

File: backend/app/schemas/analysis.py
"""PROJ-23: Schemas für Erfolgsfaktorenanalyse."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class AnalysisRunRead(BaseModel):
    id: UUID
    comparison_group: str
    success_definition: dict
    total_analyzed: int
    total_excluded: int
    excluded_reasons: dict
    created_at: datetime


class AnalysisRunRowRead(BaseModel):
    id: UUID
    strategy_name: str
    strategy_version_id: UUID | None = None
    calmar_ratio: float | None = None
    sortino_ratio: float | None = None
    trades_per_year: float | None = None
    is_success: bool
    indicators: list[str]
    indicator_count: int
    parameter_count: int
    entry_archetype: str
    exit_archetype: str
    category: str | None = None
    direction: str | None = None
    mts_compatibility: str | None = None


class CohortRow(BaseModel):
    value: str
    success: int
    total: int
    success_quote: float | None = None
    lift: float | None = None
    median_calmar: float | None = None


class AnalysisRunDetailRead(AnalysisRunRead):
    rows: list[AnalysisRunRowRead]
    cohort: list[CohortRow]
